api/lib/flight_job_script_api/job_cli.rb

```diff
@@ -0,0 +1,212 @@
+        sys.run do
+          File.write answers_path, answers if answers
+          File.write notes_path, notes if notes
+        end
+      ensure
+        FileUtils.rm_f answers_path
+        FileUtils.rm_f notes_path
+      end
+
+      def edit_script_notes(script_id, **opts)
+        new(*flight_job, 'edit-script-notes', script_id, '--json', '--notes', '@-', **opts).run
+      end
+
+      def edit_script(script_id, **opts)
+        new(*flight_job, 'edit-script', script_id, '--json', '--force', '--content', '@-', **opts).run
+      end
+
+      def delete_script(id, **opts)
+        new(*flight_job, 'delete-script', id, '--json',**opts).run
+      end
+
+      def list_jobs(**opts)
+        new(*flight_job, 'list-jobs', '--json', **opts).run
+      end
+
+      def info_job(id, **opts)
+        new(*flight_job, 'info-job', id, '--json', **opts).run
+      end
+
+      def submit_job(script_id, **opts)
+        new(*flight_job, 'submit-job', script_id, '--json', **opts).run
+      end
+
+      def view_job_stdout(job_id, **opts)
+        new(*flight_job, 'view-job-stdout', job_id, **opts).run
+      end
+
+      def view_job_stderr(job_id, **opts)
+        new(*flight_job, 'view-job-stderr', job_id, **opts).run
+      end
+
+      def view_job_results(job_id, filename, **opts)
+        new(*flight_job, 'view-job-results', job_id, filename, **opts).run
+      end
+
+      def recursive_glob_dir(dir, **opts)
+        sys = new(:noop, 'recursively glob directory', **opts)
+        sys.run do |stdout, stderr|
+          exit 20 unless Dir.exists?(dir)
+          json = Dir.glob(File.join(dir, '**/*'))
+            .map { |p| Pathname.new(p) }
+            .reject(&:directory?)
+            .reject(&:symlink?)
+            .select(&:readable?) # XXX: Non-readable files would be an odd occurrence
+            .map { |p| { file: p.to_s, size: p.size } }
+          stdout.puts(JSON.generate(json))
+        end
+      end
+
+      private
+
+      def flight_job
+        FlightJobScriptAPI.config.flight_job
+      end
+    end
+
+    def initialize(*cmd, user:, stdin: nil, env: {})
+      @cmd = cmd
+      @user = user
+      @stdin = stdin
+      @env = {
+        'PATH' => FlightJobScriptAPI.app.config.command_path,
+        'HOME' => passwd.dir,
+        'USER' => @user,
+        'LOGNAME' => @user
+      }.merge(env)
+    end
+
+    def run(&block)
+      result =
+        self.class.mutexes[@user].synchronize do
+          FlightJobScriptAPI.logger.debug("Running subprocess (#{@user}): #{stringified_cmd}")
+          sp = Subprocess.new(
+            dir: passwd.dir,
+            env: @env,
+            gid: passwd.gid,
+            logger: FlightJobScriptAPI.logger,
+            timeout: FlightJobScriptAPI.config.command_timeout,
+            uid: passwd.uid,
+          )
+          sp.run(@cmd.first == :noop ? nil : @cmd, @stdin, &block)
+        end
+      parse_result(result)
+      log_command(result)
```

Comment:
You're right that all sub commands should all be logged. Unfortunately knowing how to stringify the command is not entirely straight forward.  There may be sensitive data that we don't want to appear in the logs or for `JobCLI` there is an issue with stringifying `noop` commands.  I came to the conclusion that punting for now was the best option.

Reply:
Yeah I'm not a fan of the `noop` commands. They only exist because they where retro-fitted over the existing architecture.

What would make more sense is:

* Allow the `cmd` to be `nil` or `[]`, which flags `Subprocess` not to `exec`, and
* Add a `log_cmd` option which defaults to `cmd.join(' ')`.

This should be sufficient for the two cases.